### PR TITLE
Fix enums leq

### PR DESCRIPTION
**Pull request by jerhard**
This fixes issue #332. The implementation of `Enums.leq` used to be based on  the `join`, but this is inherently buggy for `Enums`, as the `join` does not always deliver the least upper bound of two elements. 
The new implementation has to perform a case distinction on all four possible cases. 

When checking whether an exclusion set is contained in a inclusion set or another exclusion set, we possibly have to iterate over elements in the range of the abstract value that is supposedly smaller. To avoid having to perform this check as much as possible, we first check whether the cardinality of the concretization of the first abstract value is smaller than the concretization of the second abstract value, similar to what @michael-schwarz  suggested in our discussion today.

### CODE REVIEW

**Comment by michael-schwarz** [comment]
Is there a specific reason this is still a draft? If not, I'd suggest merging it!

File: src/cdomains/intDomain.ml [hunk 1]
```diff
@@ -1314,9 +1307,9 @@ struct
     (* A definite value is leq all exclusion sets that don't contain it *)
     | `Definite x, `Excluded (s,r) -> in_range r x && not (S.mem x s)
     (* No finite exclusion set can be leq than a definite value *)
-    | `Excluded _, `Definite _ -> false
+    | `Excluded _, `Definite _ -> false (* wrong *)
     (* Excluding X <= Excluding Y whenever Y <= X *)
-    | `Excluded (x,xw), `Excluded (y,yw) -> S.subset y x && R.leq xw yw
+    | `Excluded (x,xw), `Excluded (y,yw) -> S.subset y x && R.leq xw yw (* wrong *)
```
Comment (sim642) [inline src/cdomains/intDomain.ml:1312]: I'm guessing it's a draft because of these "wrongs"?

Comment (michael-schwarz) [reply src/cdomains/intDomain.ml:1312]: Yes, good catch!

**Comment by jerhard** [comment]
I now moved the handling of comparing two Exclusion sets with `leq` into a new module `Exclusion`, as well as the handling of `leq exclusion_set inclusion_set`.

**Review by michael-schwarz — APPROVED** [review]
LGTM!

**Pull request merged**